File: plover_vm/memory/bus.py
"""Unified CPU memory bus."""

from __future__ import annotations

from plover_vm.decode import MapDecoder
from plover_vm.memory.mailbox import Mailbox
from plover_vm.memory.nor import NorFlash
from plover_vm.memory.ram import Ram64K


class MemoryBus:
    def __init__(
        self,
        nor: NorFlash | None = None,
        ram: Ram64K | None = None,
        mailbox: Mailbox | None = None,
    ) -> None:
        self.nor = nor or NorFlash()
        self.ram = ram or Ram64K()
        self.mailbox = mailbox or Mailbox()
        self.decoder = MapDecoder()
        self.map_mode = 0
        self.reset_active = False

    def fetch_addr(self, pc: int) -> int:
        if self.reset_active:
            return 0xFFFC
        return pc & 0xFFFF

    def read_cpu(self, addr: int) -> int:
        a = addr & 0xFFFF
        d = self.decoder.decode(a, self.map_mode, self.reset_active)
        if d.mailbox:
            return self.mailbox.read(a)
        if d.rom_cpu:
            return self.nor.cpu_map_read(a)
        if d.ram1 or d.ram2:
            return self.ram.read(a)
        return 0xFF

    def write_cpu(self, addr: int, val: int) -> None:
        a = addr & 0xFFFF
        d = self.decoder.decode(a, self.map_mode, self.reset_active)
        if d.mailbox:
            self.mailbox.write(a, val)
            return
        if d.rom_cpu:
            return
        if d.ram1 or d.ram2:
            self.ram.write(a, val)

File: crates/plover_mmu/src/bus.rs
use crate::decode::MapDecoder;
use crate::nor::NorFlash;
use crate::ram::Ram64K;
use plover_copro::Mailbox;

#[derive(Clone, Debug)]
pub struct MemoryBus {
    pub nor: NorFlash,
    pub ram: Ram64K,
    pub mailbox: Mailbox,
    decoder: MapDecoder,
    pub map_mode: u8,
    pub reset_active: bool,
}

impl Default for MemoryBus {
    fn default() -> Self {
        Self {
            nor: NorFlash::new(),
            ram: Ram64K::new(),
            mailbox: Mailbox::default(),
            decoder: MapDecoder,
            map_mode: 0,
            reset_active: false,
        }
    }
}

impl MemoryBus {
    pub fn fetch_addr(&self, pc: u16) -> u16 {
        if self.reset_active {
            0xFFFC
        } else {
            pc
        }
    }

    pub fn read_cpu(&self, addr: u16) -> u8 {
        let d = self.decoder.decode(addr, self.map_mode, self.reset_active);
        if d.mailbox {
            return self.mailbox.read(addr);
        }
        if d.rom_cpu {
            return self.nor.cpu_map_read(addr);
        }
        if d.ram1 || d.ram2 {
            return self.ram.read(addr);
        }
        0xFF
    }

    pub fn write_cpu(&mut self, addr: u16, val: u8) {
        let d = self.decoder.decode(addr, self.map_mode, self.reset_active);
        if d.mailbox {
            self.mailbox.write(addr, val);
            return;
        }
        if d.rom_cpu {
            return;
        }
        if d.ram1 || d.ram2 {
            self.ram.write(addr, val);
        }
    }
}
